src/build_helper/src/git.rs

```diff
@@ -173,7 +176,177 @@ pub fn get_closest_merge_commit(
     Ok(output_result(&mut git)?.trim().to_owned())
 }
 
+/// Represents the result of checking whether a set of paths
+/// have been modified locally or not.
+#[derive(PartialEq, Debug)]
+pub enum PathFreshness {
+    /// Artifacts should be downloaded from this upstream commit,
+    /// there are no local modifications.
+    LastModifiedUpstream { upstream: String },
+    /// There are local modifications to a certain set of paths.
+    /// "Local" essentially means "not-upstream" here.
+    /// `upstream` is the latest upstream merge commit that made modifications to the
+    /// set of paths.
+    HasLocalModifications { upstream: String },
+}
+
+/// This function figures out if a set of paths was last modified upstream or
+/// if there are some local modifications made to them.
+///
+/// It can be used to figure out if we should download artifacts from CI or rather
+/// build them locally.
+///
+/// `target_paths` should be a non-empty slice of paths (relative to `git_dir` or the
+/// current working directory) whose modifications would invalidate the artifact.
+/// Each path can also be a negative match, i.e. `:!foo`. This matches changes outside
+/// the `foo` directory.
+///
+/// The function behaves differently in CI and outside CI.
+///
+/// - Outside CI, we want to find out if `target_paths` were modified in some local commit on
+/// top of the local master branch.
+/// If not, we try to find the most recent upstream commit (which we assume are commits
+/// made by bors) that modified `target_paths`.
+/// We don't want to simply take the latest master commit to avoid changing the output of
+/// this function frequently after rebasing on the latest master branch even if `target_paths`
+/// were not modified upstream in the meantime. In that case we would be redownloading CI
+/// artifacts unnecessarily.
+///
+/// - In CI, we always fetch only a single parent merge commit, so we do not have access
+/// to the full git history.
+/// Luckily, we only need to distinguish between two situations. The first is that the current
+/// PR made modifications to `target_paths`. If not, then we simply take the latest upstream
+/// commit, because on CI there is no need to avoid redownloading.
```

Comment:
Nit:

> Luckily, we only need to distinguish between two situations.
>
> 1. Current PR made modifications to `target_paths`. Then a build is necessary.
> 2. Current PR made no modifications to `target_paths`. Then we simply take the latest upstream commit, because on CI there is no need to avoid redownloading.